###MultiCollinearity in Regression

--> Here we find the collinearity between the features and their collinearity percentages.

--> Collinearity is the percentage of features which will depend on other which we are removed so that it will not impact the model.

--> We add constant to the model and find the summary of the model such that the standarderror is seen and P value is observed such that the standarderror should be low and corresponding P value should less than the 0.5 so that the conclusion that the dependent on the other feature is low.

In [18]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [19]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [20]:
df=pd.read_csv('/content/drive/MyDrive/Classroom/Salary_Data.csv')
df

YearsExperience   Age  Salary
0               1.1  21.0   39343
1               1.3  21.5   46205
2               1.5  21.7   37731
3               2.0  22.0   43525
4               2.2  22.2   39891
5               2.9  23.0   56642
6               3.0  23.0   60150
7               3.2  23.3   54445
8               3.2  23.3   64445
9               3.7  23.6   57189
10              3.9  23.9   63218
11              4.0  24.0   55794
12              4.0  24.0   56957
13              4.1  24.0   57081
14              4.5  25.0   61111
15              4.9  25.0   67938
16              5.1  26.0   66029
17              5.3  27.0   83088
18              5.9  28.0   81363
19              6.0  29.0   93940
20              6.8  30.0   91738
21              7.1  30.0   98273
22              7.9  31.0  101302
23              8.2  32.0  113812
24              8.7  33.0  109431
25              9.0  34.0  105582
26              9.5  35.0  116969
27              9.6  36.0  112635
28             10.3  37.0  122391
29             10.5  38.0  121872

In [21]:
import statsmodels.api as sm

In [22]:
x=df[['YearsExperience','Age']]
y=df['Salary']

In [23]:
x

YearsExperience   Age
0               1.1  21.0
1               1.3  21.5
2               1.5  21.7
3               2.0  22.0
4               2.2  22.2
5               2.9  23.0
6               3.0  23.0
7               3.2  23.3
8               3.2  23.3
9               3.7  23.6
10              3.9  23.9
11              4.0  24.0
12              4.0  24.0
13              4.1  24.0
14              4.5  25.0
15              4.9  25.0
16              5.1  26.0
17              5.3  27.0
18              5.9  28.0
19              6.0  29.0
20              6.8  30.0
21              7.1  30.0
22              7.9  31.0
23              8.2  32.0
24              8.7  33.0
25              9.0  34.0
26              9.5  35.0
27              9.6  36.0
28             10.3  37.0
29             10.5  38.0

In [24]:
x=sm.add_constant(x)  # Here we added one column constant such that all  values are 1.0
x

const  YearsExperience   Age
0     1.0              1.1  21.0
1     1.0              1.3  21.5
2     1.0              1.5  21.7
3     1.0              2.0  22.0
4     1.0              2.2  22.2
5     1.0              2.9  23.0
6     1.0              3.0  23.0
7     1.0              3.2  23.3
8     1.0              3.2  23.3
9     1.0              3.7  23.6
10    1.0              3.9  23.9
11    1.0              4.0  24.0
12    1.0              4.0  24.0
13    1.0              4.1  24.0
14    1.0              4.5  25.0
15    1.0              4.9  25.0
16    1.0              5.1  26.0
17    1.0              5.3  27.0
18    1.0              5.9  28.0
19    1.0              6.0  29.0
20    1.0              6.8  30.0
21    1.0              7.1  30.0
22    1.0              7.9  31.0
23    1.0              8.2  32.0
24    1.0              8.7  33.0
25    1.0              9.0  34.0
26    1.0              9.5  35.0
27    1.0              9.6  36.0
28    1.0             10.3  37.0
29    1.0             10.5  38.0

In [25]:
model=sm.OLS(y,x).fit() # Here first value we need to pass y and then x and fit()

In [26]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Salary   R-squared:                       0.960
Model:                            OLS   Adj. R-squared:                  0.957
Method:                 Least Squares   F-statistic:                     323.9
Date:                Tue, 10 Dec 2024   Prob (F-statistic):           1.35e-19
Time:                        09:55:46   Log-Likelihood:                -300.35
No. Observations:                  30   AIC:                             606.7
Df Residuals:                      27   BIC:                             610.9
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const           -6661.9872   2.28e+04     -0.292      0.773   -5.35e+04    4.02e+04
YearsExperience  6153.3533   2337.092      2.633      0.014    1358.037    1.09e+04
Age              1836.0136   1285.034      1.429      0.165    -800.659    4472.686
==============================================================================
Omnibus:                        2.695   Durbin-Watson:                   1.711
Prob(Omnibus):                  0.260   Jarque-Bera (JB):                1.975
Skew:                           0.456   Prob(JB):                        0.372
Kurtosis:                       2.135   Cond. No.                         626.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Here first we see r2 that is fine

Now we go to standard error is too high so that we may assume that it has some collinearity between the features.

Now we go to p value for age the p values is high greater than of .5.

So if we want to remove one feature we  can remove the high p value such that we can overcome  collinearity between them.

Further we want to check collinearity we use collinearity matrix.

In [27]:
x.iloc[:,1:].corr()

YearsExperience       Age
YearsExperience         1.000000  0.987258
Age                     0.987258  1.000000

###conclusion

Here we can see that 98% corelation between the features we can drop the age column such that it contains high p value to optimize the model.